beacon-chain/forkchoice/protoarray/optimistic_sync.go

```diff
@@ -0,0 +1,79 @@
+package protoarray
+
+import (
+	"fmt"
+
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/config/params"
+	"github.com/prysmaticlabs/prysm/encoding/bytesutil"
+)
+
+// This returns the minimum and maximum slot of the synced_tips tree
+func (f *ForkChoice) boundarySyncedTips() (types.Slot, types.Slot) {
+	f.syncedTips.RLock()
+	defer f.syncedTips.RUnlock()
+
+	min := params.BeaconConfig().FarFutureSlot
+	max := types.Slot(0)
+	for _, slot := range f.syncedTips.validatedTips {
+		if slot > max {
+			max = slot
+		}
+		if slot < min {
+			min = slot
+		}
+	}
+	return min, max
+}
+
+// Optimistic returns true if this node is optimistically synced
+// WARNING: this function does not check if slot corresponds to the
+//          block with the given root. The caller is responsible for
+//          checking this. A consensus bug could be a cause of getting
+//          this wrong, so think twice before passing a wrong pair.
+func (f *ForkChoice) Optimistic(root [32]byte, slot types.Slot) (bool, error) {
+	// If the node is a synced tip, then it's fully validated
+	f.syncedTips.RLock()
+	_, ok := f.syncedTips.validatedTips[root]
+	if ok {
+		return false, nil
+	}
+	f.syncedTips.RUnlock()
+
+	// If the slot is higher than the max synced tip, it's optimistic
+	min, max := f.boundarySyncedTips()
+	if slot > max {
+		return true, nil
+	}
+
+	// If the slot is lower than the min synced tip, it's fully validated
+	if slot <= min {
+		return false, nil
+	}
+
+	// If we reached this point then the block has to be in the Fork Choice
+	// Store!
+	f.store.nodesLock.RLock()
+	index, ok := f.store.nodesIndices[root]
```

Comment:
defer f.store.nodesLock.RUnlock() is a better pattern.

Reply:
I cannot do this in a recursive function